spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,292 @@
+  private Offset endOffset;
+
+  private final Table table;
+  private final int maxFilesPerTrigger;
+  private final long splitSize;
+  private final int lookback;
+  private final long openFileCost;
+  private final Offset initialOffset;
+
+  private final Cache<Long, CloseableIterable<FileScanTask>> snapshotTasksCache = Caffeine.newBuilder()
+      .maximumSize(CACHE_SIZE)
+      .softValues()
+      .build();
+
+  /**
+   * Utility class to track the snapshotId, task and index of task within this snapshot.
+   */
+  @VisibleForTesting
+  static class IndexedTask {
+    private final long snapshotId;
+    private final FileScanTask task;
+    private final int index;
+
+    IndexedTask(long snapshotId, FileScanTask task, int index) {
+      this.snapshotId = snapshotId;
+      this.task = task;
+      this.index = index;
+    }
+
+    long snapshotId() {
+      return snapshotId;
+    }
+
+    FileScanTask task() {
+      return task;
+    }
+
+    int index() {
+      return index;
+    }
+  }
+
+  StreamingReader(Table table, Broadcast<FileIO> io, Broadcast<EncryptionManager> encryptionManager,
+                  boolean caseSensitive, DataSourceOptions options) {
+    super(table, io, encryptionManager, caseSensitive, options);
+
+    this.table = table;
+    this.maxFilesPerTrigger =
+      options.get("max-files-per-trigger").map(Integer::parseInt).orElse(DEFAULT_MAX_FILES_PER_TRIGGER);
+    Preconditions.checkArgument(maxFilesPerTrigger > 0,
+        "Option max-files-per-trigger '%d' should > 0", maxFilesPerTrigger);
+
+    Long snapshotId = options.get("snapshot-id").map(Long::parseLong).orElse(null);
+    if (snapshotId != null) {
+      if (SnapshotUtil.ancestorOf(table, table.currentSnapshot().snapshotId(), snapshotId)) {
+        this.initialOffset = new StreamingOffset(snapshotId, -1, true);
+      } else {
+        LOG.warn("The option snapshot-id {} is not a valid snapshot id, will use the earliest " +
+            "snapshot instead", snapshotId);
+        this.initialOffset = calculateInitialOffset();
+      }
+    } else {
+      this.initialOffset = calculateInitialOffset();
+    }
+
+    this.splitSize = Optional.ofNullable(table.properties().get(TableProperties.SPLIT_SIZE))
+        .map(Long::parseLong)
+        .orElse(TableProperties.SPLIT_SIZE_DEFAULT);
+    this.lookback = Optional.ofNullable(table.properties().get(TableProperties.SPLIT_LOOKBACK))
+        .map(Integer::parseInt)
+        .orElse(TableProperties.SPLIT_LOOKBACK_DEFAULT);
+    this.openFileCost = Optional.ofNullable(table.properties().get(TableProperties.SPLIT_OPEN_FILE_COST))
+        .map(Long::parseLong)
+        .orElse(TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT);
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public void setOffsetRange(Optional<Offset> start, Optional<Offset> end) {
+    table.refresh();
+
+    this.startOffset = start.orElse(initialOffset);
+    if (startOffset.equals(StreamingOffset.INVALID_OFFSET)) {
```

Comment:
If there are no snapshots, then the initial offset will be `StreamingOffset.INVALID_OFFSET`. In that case, the reader never recovers and makes progress because this always sets the end offset to `StreamingOffset.INVALID_OFFSET`.

Instead, I think this should rename `INVALID_OFFSET` to `START_OFFSET` and use it as the only offset where `is_starting` is true. That way, if `is_starting` is set, then the snapshot ID and index are ignored (as they are not known). When calculating the end offset, you can always use the most recent snapshot and index.

I think that would allow you to start a streaming job on a table with no data and have it make progress when the first commit is added.

Reply:
Yeah, this indeed is an issue, let me fix it.